File: Projects/Multithreading/app/src/main/java/com/ivettevaldez/multithreading/demos/demorxjava/ProducerConsumerBenchmarkUseCase.java
package com.ivettevaldez.multithreading.demos.demorxjava;

import android.util.Log;

import java.util.concurrent.atomic.AtomicInteger;

import io.reactivex.Observable;

class ProducerConsumerBenchmarkUseCase {

    private final static int NUM_OF_MESSAGES = 1000;
    private final static int BLOCKING_QUEUE_CAPACITY = 5;
    private final static Object LOCK = new Object();

    private final String classTag = this.getClass().getSimpleName();
    private final MyBlockingQueue blockingQueue = new MyBlockingQueue(BLOCKING_QUEUE_CAPACITY);
    private final AtomicInteger numOfThreads = new AtomicInteger(0);

    private int numOfReceivedMessages;
    private int numOfFinishedConsumers;
    private long startTimestamp;

    Observable<Result> startBenchmark() {
        return Observable.fromCallable(
                () -> {
                    synchronized (LOCK) {
                        numOfReceivedMessages = 0;
                        numOfFinishedConsumers = 0;
                        numOfThreads.set(0);
                        startTimestamp = System.currentTimeMillis();
                    }

                    // Producers init thread.
                    new Thread(() -> {
                        for (int i = 0; i < NUM_OF_MESSAGES; i++) {
                            startNewProducer(i);
                        }
                    }).start();

                    // Consumers init thread.
                    new Thread(() -> {
                        for (int i = 0; i < NUM_OF_MESSAGES; i++) {
                            startNewConsumer();
                        }
                    }).start();

                    synchronized (LOCK) {
                        while (numOfFinishedConsumers < NUM_OF_MESSAGES) {
                            try {
                                LOCK.wait();
                            } catch (InterruptedException ex) {
                                Log.e(classTag, ex.toString());
                                return new Result(
                                        System.currentTimeMillis() - startTimestamp,
                                        -1
                                );
                            }
                        }

                        return new Result(
                                System.currentTimeMillis() - startTimestamp,
                                numOfReceivedMessages
                        );
                    }
                }
        );
    }

    private void startNewProducer(final int index) {
        new Thread(() -> blockingQueue.put(index)).start();
    }

    private void startNewConsumer() {
        new Thread(() -> {
            int message = blockingQueue.take();

            synchronized (LOCK) {
                if (message != -1) {
                    numOfReceivedMessages++;
                }
                numOfFinishedConsumers++;
                LOCK.notifyAll();
            }
        }).start();
    }

    static class Result {

        private final long executionTime;
        private final int numberOfMessages;

        Result(long executionTime, int numberOfMessages) {
            this.executionTime = executionTime;
            this.numberOfMessages = numberOfMessages;
        }

        long getExecutionTime() {
            return executionTime;
        }

        int getNumberOfMessages() {
            return numberOfMessages;
        }
    }
}

File: Courses/Multithreading/app/src/main/java/com/ivettevaldez/multithreading/demos/democoroutines/ProducerConsumerBenchmarkUseCase.kt
package com.ivettevaldez.multithreading.demos.democoroutines

import android.util.Log
import kotlinx.coroutines.*
import java.util.concurrent.atomic.AtomicInteger

private const val NUM_OF_MESSAGES = 1000
private const val BLOCKING_QUEUE_CAPACITY = 5

class ProducerConsumerBenchmarkUseCase {

    private val classTag: String = this.javaClass.simpleName
    private val blockingQueue = MyBlockingQueue(BLOCKING_QUEUE_CAPACITY)

    private var numOfReceivedMessages: AtomicInteger = AtomicInteger(0)
    private var numOfConsumers: AtomicInteger = AtomicInteger(0)

    @Volatile
    private var startTimestamp: Long = 0

    class Result(val executionTime: Long, val numberOfMessages: Int)

    suspend fun startBenchmark(): Result {
        withContext(Dispatchers.IO) {
            numOfReceivedMessages.set(0)
            startTimestamp = System.currentTimeMillis()

            // Producers init coroutine.
            val deferredProducers = async(Dispatchers.IO + NonCancellable) {
                for (i in 0 until NUM_OF_MESSAGES) {
                    startNewProducer(i)
                }
            }

            // Consumers init coroutine.
            val deferredConsumers = async(Dispatchers.IO + NonCancellable) {
                for (i in 0 until NUM_OF_MESSAGES) {
                    startNewConsumer()
                }
            }

            awaitAll(deferredProducers, deferredConsumers)
        }

        return Result(
                System.currentTimeMillis() - startTimestamp,
                numOfReceivedMessages.get()
        )
    }

    private fun CoroutineScope.startNewProducer(i: Int) = launch(Dispatchers.IO) {
        Log.d(classTag, "Producer $i started on Thread ${Thread.currentThread().name}")
        blockingQueue.put(i)
    }

    private fun CoroutineScope.startNewConsumer() = launch(Dispatchers.IO) {
        val message = blockingQueue.take()
        if (message != -1) {
            numOfReceivedMessages.incrementAndGet()
        }
        Log.d(classTag,
                "Consumer ${numOfConsumers.incrementAndGet()} " +
                        "started on Thread ${Thread.currentThread().name}"
        )
    }
}
